### PR TITLE
Spellbook refactor

### PR DESCRIPTION
**Pull request by unid15**
About time this was done

- Rewrote spellbook code to be great and modular
- Added separate buttons for upgrading spells. Clicking on the upgrade name (cooldown, power) will tell you what upgrading the spell will do
- Bought spells are shown on top of the list
- This also makes changing spell prices very possible!

TODO:
- [x] Fix some broken shit
- [x] Re-add messages on purchasing spells/artifacts
- [x] Fix broken descriptions and some formatting issues
- [x] Test thoroughly
- [x] Make it prettier, maybe? Spell icons or something fancy like a background color.
- [x] Re-implement feedback

~~Question:~~

~~How important is feedback? This stuff: feedback_add_details("wizard_spell_learned","WM") //please do not change the abbreviation to keep data processing consistent. Add a unique id to any new spells~~

### CODE REVIEW

File: code/game/gamemodes/wizard/spellbook.dm [hunk 1]
```diff
@@ -24,383 +74,214 @@
...
+
+			//<b>Fireball</b> - 10 seconds (buy for 1 spell point)
+			//<i>(Description)</i>
+			//Requires robes to cast
+
+			var/spell_name = initial(abstract_spell.name)
+			var/spell_cooldown = get_spell_cooldown_string(initial(abstract_spell.charge_max), initial(abstract_spell.charge_type))
+			var/spell_price = get_spell_price(abstract_spell)
+
+			dat += "<b>[spell_name]</b> - [spell_cooldown] ([buy_href_link(spell_path, spell_price, "buy for [spell_price] point\s")])<br>"
```
Comment (ComicIronic) [inline code/game/gamemodes/wizard/spellbook.dm]: <b> is deprecated, and <i> is no longer directly italics in HTML5: instead, you should use <strong> and <em> respectively.

File: code/game/gamemodes/wizard/artifacts.dm [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	desc = "An artefact that spits bolts of coruscating energy which cause the target's very form to reshape itself."
+	spawned_items = list(/obj/item/weapon/gun/energy/staff)
+
+/datum/spellbook_artifact/mental_focus
+	name = "Mental Focus"
+	desc = "An artefact that channels the will of the user into destructive bolts of force."
+	spawned_items = list(/obj/item/weapon/gun/energy/staff/focus)
+
+/datum/spellbook_artifact/shards
+	name = "Six Soul Stone Shards"
```
Comment (ComicIronic) [inline code/game/gamemodes/wizard/artifacts.dm]: Should probably mention the belt as well.

**Comment by gbasood** [comment]
Feedback is the statistics gathering that pomf wrote that goes directly to SQL and is as far as I know unused
As long as statistics gathering via statisticscollection.dm isn't broken then :+1:

**Comment by unid15** [comment]
Ready for review

File: code/game/gamemodes/wizard/spellbook.dm [hunk 3]
```diff
@@ -7,10 +9,58 @@
...
+	)
+
+	//Unlike the list above, the available_artifacts list builds itself from all subtypes of /datum/spellbook_artifact
+	var/list/available_artifacts = list()
+
+	var/uses = STARTING_USES
+	var/max_uses = STARTING_USES
+
 	var/op = 1
+	var/temp = null
```
Comment (ComicIronic) [inline code/game/gamemodes/wizard/spellbook.dm]: Since it looks like you're not using this for anything anymore, you should remove it.

**Comment by ghost** [comment]
@PJB3005 /vg/UI when

File: code/game/gamemodes/wizard/artifacts.dm [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	H.add_spell(new/spell/aoe_turf/conjure/snowmobile)
+	H.add_spell(new/spell/targeted/wrapping_paper)
+	H.add_spell(new/spell/aoe_turf/conjure/gingerbreadman)
+	H.add_spell(new/spell/targeted/flesh_to_coal)
+
+	to_chat(world,'sound/misc/santa.ogg')
+	SetUniversalState(/datum/universal_state/christmas)
+
+/datum/spellbook_artifact/santa_bundle/can_buy()
+	return (Holiday == "Christmas" && universe.name == "Normal")
```
Comment (ghost) [inline code/game/gamemodes/wizard/artifacts.dm]: Universe should be an istype

File: code/game/gamemodes/wizard/spellbook.dm [hunk 5]
```diff
@@ -24,383 +73,241 @@
...
+	if(uses >= amount)
+		uses -= amount
+		if(no_refunds)
+			max_uses -= amount
+
 		return 1
 
-	if(H.mind.special_role == "apprentice")
-		temp = "If you got caught sneaking a peak from your teacher's spellbook, you'd likely be expelled from the Wizard Academy. Better not."
+	return 0
```
Comment (ghost) [inline code/game/gamemodes/wizard/spellbook.dm]: Trailing return

File: code/game/gamemodes/wizard/spellbook.dm [hunk 6]
```diff
@@ -24,383 +73,241 @@
...
+
+/obj/item/weapon/spellbook/proc/use(amount, no_refunds = 0)
+	if(uses >= amount)
+		uses -= amount
+		if(no_refunds)
+			max_uses -= amount
+
 		return 1
 
-	if(H.mind.special_role == "apprentice")
```
Comment (ghost) [inline code/game/gamemodes/wizard/spellbook.dm]: Was this moved elsewhere?

**Comment by unid15** [comment]
Conflicts fixed

**Comment by unid15** [comment]
(they are)

**Comment by ComicIronic** [comment]
Let's see what this breaks.

**Comment by clusterfack** [comment]
Pretty sure this now lets you refund the spells of others who have used the book before you. Sounds like a raging mages disaster waiting to happen.

**Pull request merged**